src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt

```diff
@@ -0,0 +1,332 @@
+        }
+
+        val fc = JFileChooser()
+        fc.addChoosableFileFilter(PNGFileFilter())
+        fc.isAcceptAllFileFilterUsed = false
+
+        JButton("Load Color Map").also {
+            it.addActionListener {
+                val returnVal: Int = fc.showOpenDialog(this)
+                if (returnVal == JFileChooser.APPROVE_OPTION) loadFromFile(fc.selectedFile)
+            }
+            add(it)
+        }
+
+        JButton("Save Color Map").also {
+            it.addActionListener {
+                val option = fc.showSaveDialog(this)
+                if (option == JFileChooser.APPROVE_OPTION) {
+                    saveToFile (fc.selectedFile)
+                }
+            }
+            add(it,"wrap")
+        }
+    }
+
+    /**
+     * A filter to only select pngs.
+     */
+    class PNGFileFilter: FileFilter()
+    {
+        override fun accept(f: File): Boolean {
+            if (f.isDirectory()) {
+                return false
+            }
+            if (f.extension != "png") return false
+            return true
+        }
+
+        override fun getDescription(): String {
+            return "png"
+        }
+    }
+
+    /**
+     * Save a color map to file as a png.
+     */
+    fun saveToFile(file: File){
+        var file_ = file
+        if (FilenameUtils.getExtension(file.name).equals("png", ignoreCase = true)) {
+            // filename is OK as-is
+        } else {
+            file_ = File("$file.png") // append .png if "foo.jpg.png" is OK
+        }
+
+        try {
+            ImageIO.write(colorMapEditor.toImage(), "png", file_)
+        } catch (ioe: IOException) {
+            ioe.printStackTrace()
+        }
+    }
+
+    /**
+     * Load a color map from png.
+     */
+    fun loadFromFile(file: File){
+        var img: BufferedImage? = null
+        try {
+            img = ImageIO.read(file)
+        } catch (_: IllegalArgumentException){
+            logger.error("Could not find file ${file.path}")
+        } catch (e: IOException){
+            logger.error(e.toString())
+        }
+        if (img == null) return
+        val cm = Colormap.fromBuffer(Image.bufferedImageToRGBABuffer(img),img.width, img.height)
+        colorMapEditor.loadColormap(cm)
+    }
+}
+
+/**
+ * A Gui element to allow users to visually create or modify a color map
+ */
+class ColorMapEditor(var target:Volume? = null) : JPanel() {
+
+    private var colorPoints = listOf(
+        ColorPoint(0.0f, Color(0f, 0f, 0f)),
+        ColorPoint(0.5f, Color(0f, 0.5f, 0f)),
+        ColorPoint(1f, Color(0f, 1f, 0f))
+    )
+
+    private var hoveredOver: ColorPoint? = null
+    private var dragging: ColorPoint? = null
+    private var dragged = false
+
+    init {
+        this.layout = MigLayout()
+        this.preferredSize = Dimension(1000, 40)
+
+        target?.let { loadColormap(it.colormap) }
+
+        this.addMouseListener(object : MouseListener {
+            override fun mouseClicked(e: MouseEvent) {
```

Comment:
Missing docs

Reply:
I am a bit confused about the standart of documentation we are aming for. We are not enforcing documentation on private methods (which is the correct way to go IMO) then why are we enforcing documentation on an overwrite of an GUI Listener interface?